front/app/containers/IdeasShow/GoBackButton.tsx

```diff
@@ -1,61 +1,92 @@
 import React, { memo } from 'react';
-import clHistory from 'utils/cl-router/history';
+
+// components
 import { Button, useBreakpoint, Box } from '@citizenlab/cl2-component-library';
+
+// hooks
 import useProject from 'hooks/useProject';
 import useLocale from 'hooks/useLocale';
-import useLocalize from 'hooks/useLocalize';
+
+// utils
+import clHistory from 'utils/cl-router/history';
 import { isNilOrError } from 'utils/helperUtils';
 import eventEmitter from 'utils/eventEmitter';
 import { ScreenReaderOnly } from 'utils/a11y';
 
+// i18n
+import messages from './messages';
+import { injectIntl, FormattedMessage } from 'utils/cl-intl';
+import { InjectedIntlProps } from 'react-intl';
+
 interface Props {
   projectId: string;
   className?: string;
   insideModal: boolean;
+  deselectIdeaOnMap?: () => void;
 }
 
-const GoBackButton = memo(({ projectId, className, insideModal }: Props) => {
-  const project = useProject({ projectId });
-  const locale = useLocale();
-  const isSmallTablet = useBreakpoint('smallTablet');
-  const localize = useLocalize();
-
-  const onGoBack = (event: React.MouseEvent) => {
-    event.preventDefault();
-
-    if (insideModal) {
-      eventEmitter.emit('closeIdeaModal');
-    } else if (!isNilOrError(project)) {
-      clHistory.push(`/projects/${project.attributes.slug}`);
-    } else {
-      clHistory.push('/');
+const GoBackButton = memo(
+  ({
+    projectId,
+    className,
+    insideModal,
+    deselectIdeaOnMap,
+  }: Props & InjectedIntlProps) => {
+    const project = useProject({ projectId });
+    const locale = useLocale();
+    const isSmallTablet = useBreakpoint('smallTablet');
+
+    const projectExists = !isNilOrError(project);
+    const deselectIdeaCallbackExists = !isNilOrError(deselectIdeaOnMap);
+
+    // a "back to ideas" message when returning to the list of ideas (e.g. on map)
+    // or a generic "go back message" when returning to the project main page
+    const buttonMessage =
+      projectExists && deselectIdeaCallbackExists
+        ? messages.backToIdeas
+        : messages.goBack;
+
+    const onGoBack = (event: React.MouseEvent) => {
+      event.preventDefault();
+
+      if (insideModal) {
+        eventEmitter.emit('closeIdeaModal');
+      } else if (projectExists && deselectIdeaCallbackExists) {
+        // if deselectIdeaOnMap is present, call it to switch view on the map overlay
+        deselectIdeaOnMap();
+      } else if (projectExists) {
+        // if deselectIdeaOnMap is not present, link back to main project
+        clHistory.push(`/projects/${project.attributes.slug}`);
```

Comment:
I find it easier to reason about the code if we just use `if`/`else` and not add `else if`. I also find it harder to understand the code if the same condition (`projectExists`) is repeated in different if blocks. And it might be easier to introduce a bug. What do you think?

```
if (insideModal) {
  eventEmitter.emit('closeIdeaModal');
} else {
  if (projectExists) {
    if (deselectIdeaCallbackExists) {
      // if deselectIdeaOnMap is present, call it to switch view on the map overlay
      deselectIdeaOnMap();
    } else {
      // if deselectIdeaOnMap is not present, link back to main project
      clHistory.push(`/projects/${project.attributes.slug}`);
    }
  } else {
    clHistory.push('/');
  }
}
```

Reply:
I'm usually hesitant to nesting conditionals like that, IMO that can be more difficult to follow the logic. Maybe something like one of these?

 ```suggestion
    const onGoBack = (event: React.MouseEvent) => {
      event.preventDefault();

      if (insideModal) {
        eventEmitter.emit('closeIdeaModal');
        return;
      }
      
      if (projectExists && deselectIdeaCallbackExists) {
        deselectIdeaOnMap();
        return;
      }

      if (projectExists) {
        // if deselectIdeaOnMap is not present, link back to main project
        clHistory.push(`/projects/${project.attributes.slug}`);
        return;
      }
      
      clHistory.push('/');
    };
```

or 

```suggestion
      const onGoBack = (event: React.MouseEvent) => {
      event.preventDefault();

      if (insideModal) {
        eventEmitter.emit('closeIdeaModal');
        return;
      }
      
      if (projectExists) {
        if (deselectIdeaCallbackExists) {
          deselectIdeaOnMap();
          return
        }

        // if deselectIdeaOnMap is not present, link back to main project
        clHistory.push(`/projects/${project.attributes.slug}`);
        return;
      }

      clHistory.push('/');
    };
```

Reply:
Yeah, that looks better imo. Slight preference for the 2nd option but both look good :-)

Reply:
Cool, let's go with door number two then 🥈